dbcon/joblist/tuplehashjoin.cpp

```diff
@@ -200,140 +202,237 @@ void TupleHashJoinStep::join()
     }
 }
 
-/* Index is which small input to read. */
-void TupleHashJoinStep::smallRunnerFcn(uint32_t index)
+// simple sol'n.  Poll mem usage of Joiner once per second.  Request mem
+// increase after the fact.  Failure to get mem will be detected and handled by
+// the threads inserting into Joiner.
+void TupleHashJoinStep::trackMem(uint index)
 {
-    uint64_t i;
-    bool more, flippedUMSwitch = false, gotMem;
-    RGData oneRG;
-    //shared_array<uint8_t> oneRG;
-    Row r;
-    JoinType jt;
-    RowGroupDL* smallDL;
-    uint32_t smallIt;
-    RowGroup smallRG;
-    boost::shared_ptr<TupleJoiner> joiner;
+    boost::shared_ptr<TupleJoiner> joiner = joiners[index];
+    ssize_t memBefore = 0, memAfter = 0;
+    bool gotMem;
+
+    boost::unique_lock<boost::mutex> scoped(memTrackMutex);
```

Comment:
DO we really need boost for both shared and unique ptr types?

Reply:
:D  Technically no...